File: languages/php/Bloock/Client/AuthenticityClient.php
<?php

namespace Bloock\Client;

use Bloock\Bridge\Bridge;
use Bloock\Config\Config;
use Bloock\ConfigData;
use Bloock\Entity\Authenticity\Signature;
use Bloock\Entity\Authenticity\Signer;
use Bloock\Entity\Key\EcdsaKeyPair;
use Bloock\Entity\Record\Record;
use Bloock\GenerateLocalKeyRequest;
use Bloock\GetSignaturesRequest;
use Bloock\KeyType;
use Bloock\SignatureCommonNameRequest;
use Bloock\SignRequest;
use Bloock\VerifyRequest;
use Exception;

/**
 * Represents a client for interacting with the [Bloock Authenticity service](https://dashboard.bloock.com/login).
 */
class AuthenticityClient
{
    private $bridge;
    private $config;

    /**
     * Creates a new instance of the AuthenticityClient with default configuration.
     * @param ConfigData|null $config
     */
    public function __construct(ConfigData $config = null)
    {
        $this->bridge = new Bridge();
        if ($config != null) {
            $this->config = Config::newConfigData($config);
        } else {
            $this->config = Config::newConfigDataDefault();
        }
    }

    /**
     * Generates ECDSA key pair for signing records.
     * @deprecated Will be deleted in future versions. Use KeyClient.newLocalKey function instead.
     */
    public function generateEcdsaKeyPair(): EcdsaKeyPair
    {
        $req = new GenerateLocalKeyRequest();
        $req->setConfigData($this->config);
        $req->setKeyType(KeyType::EcP256k);

        $res = $this->bridge->key->GenerateLocalKey($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return EcdsaKeyPair::fromProto($res);
    }

    /**
     * Signs a Bloock record using the specified signer.
     * @param Record $record
     * @param Signer $signer
     * @return Signature
     * @throws Exception
     */
    public function sign(Record $record, Signer $signer): Signature
    {
        $req = new SignRequest();
        $req->setConfigData($this->config)->setRecord($record->toProto())->setSigner($signer->toProto());

        $res = $this->bridge->authenticity->Sign($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Signature::fromProto($res->getSignature());
    }

    /**
     * Verifies the authenticity of a Bloock record.
     * @param Record $record
     * @return bool
     * @throws Exception
     */
    public function verify(Record $record): bool
    {
        $req = new VerifyRequest();
        $req->setConfigData($this->config)->setRecord($record->toProto());

        $res = $this->bridge->authenticity->Verify($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getValid();
    }

    /**
     * Gets the signatures associated with a Bloock record.
     * @param Record $record
     * @return array
     * @throws Exception
     */
    public function getSignatures(Record $record): array
    {
        $req = new GetSignaturesRequest();
        $req->setConfigData($this->config)->setRecord($record->toProto());

        $res = $this->bridge->authenticity->GetSignatures($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        $signatures = [];
        foreach ($res->getSignatures() as $signature) {
            $signatures[] = Signature::fromProto($signature);
        }
        return $signatures;
    }
}

File: languages/java/src/main/java/com/bloock/sdk/client/AuthenticityClient.java
package com.bloock.sdk.client;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.BloockAuthenticity.*;
import com.bloock.sdk.bridge.proto.BloockConfig.ConfigData;
import com.bloock.sdk.bridge.proto.BloockKeys;
import com.bloock.sdk.bridge.proto.BloockShared.Error;
import com.bloock.sdk.config.Config;
import com.bloock.sdk.entity.authenticity.Signature;
import com.bloock.sdk.entity.authenticity.Signer;
import com.bloock.sdk.entity.key.EcdsaKeyPair;
import com.bloock.sdk.entity.key.KeyType;
import com.bloock.sdk.entity.record.Record;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Represents a client for interacting with the
 * <a href="https://dashboard.bloock.com/login">Bloock Authenticity service</a>.
 */
public class AuthenticityClient {
  private final Bridge bridge;
  private final ConfigData configData;

  /**
   * Creates a new instance of the AuthenticityClient with default configuration.
   */
  public AuthenticityClient() {
    this.bridge = new Bridge();
    this.configData = Config.newConfigDataDefault();
  }

  /**
   * Creates a new instance of the AuthenticityClient with the provided
   * configuration.
   * 
   * @param configData
   */
  public AuthenticityClient(ConfigData configData) {
    this.bridge = new Bridge();
    this.configData = Config.newConfigData(configData);
  }

  /**
   * Generates ECDSA key pair for signing records.
   * 
   * @deprecated Will be deleted in future versions. Use KeyClient.newLocalKey
   *             function instead.
   */
  @Deprecated
  public EcdsaKeyPair generateEcdsaKeyPair() throws Exception {
    BloockKeys.GenerateLocalKeyRequest request = BloockKeys.GenerateLocalKeyRequest.newBuilder()
        .setConfigData(this.configData)
        .setKeyType(KeyType.EcP256k.toProto())
        .build();

    BloockKeys.GenerateLocalKeyResponse response = this.bridge.getKey().generateLocalKey(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return EcdsaKeyPair.fromProto(response);
  }

  /**
   * Signs a Bloock record using the specified signer.
   * 
   * @param record
   * @param signer
   * @return
   * @throws Exception
   */
  public Signature sign(Record record, Signer signer) throws Exception {
    SignRequest request = SignRequest.newBuilder()
        .setConfigData(this.configData)
        .setRecord(record.toProto())
        .setSigner(signer.toProto())
        .build();
    SignResponse response = this.bridge.getAuthenticity().sign(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Signature.fromProto(response.getSignature());
  }

  /**
   * Verifies the authenticity of a Bloock record.
   * 
   * @param record
   * @return
   * @throws Exception
   */
  public boolean verify(Record record) throws Exception {
    VerifyRequest request = VerifyRequest.newBuilder()
        .setConfigData(this.configData)
        .setRecord(record.toProto())
        .build();
    VerifyResponse response = this.bridge.getAuthenticity().verify(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getValid();
  }

  /**
   * Gets the signatures associated with a Bloock record.
   * 
   * @param record
   * @return
   * @throws Exception
   */
  public List<Signature> getSignatures(Record record) throws Exception {
    GetSignaturesRequest req = GetSignaturesRequest.newBuilder()
        .setConfigData(this.configData)
        .setRecord(record.toProto())
        .build();
    GetSignaturesResponse recordSignatures = this.bridge.getAuthenticity().getSignatures(req);

    if (recordSignatures.getError() != Error.getDefaultInstance()) {
      throw new Exception(recordSignatures.getError().getMessage());
    }

    return recordSignatures.getSignaturesList().stream()
        .map(x -> Signature.fromProto(x))
        .collect(Collectors.toList());
  }
}
